File: frontend/src/services/reports.ts
/**
 * Reports API Service
 */
import { http } from './http';

export type StockStatus = 'in_stock' | 'low_stock' | 'out_of_stock';
export type StockTransactionType = 'in' | 'out' | 'adjust' | 'transfer';

export interface InventoryReportItem {
    product_id: number;
    sku: string;
    name: string;
    category: string | null;
    quantity: number;
    unit: string | null;
    cost_price: number;
    selling_price: number;
    stock_value: number;
    stock_status: string;
    reorder_level: number;
}

export interface SalesReportItem {
    sale_id: number;
    invoice_number: string;
    sale_date: string;
    customer_id: number | null;
    customer_name: string | null;
    total_amount: number;
    discount_amount: number;
    tax_amount: number;
    status: string;
    payment_method: string | null;
}

export interface SalesReportSummary {
    total_sales: number;
    total_revenue: number;
    avg_order_value: number;
    total_discounts: number;
}

export interface PurchaseReportItem {
    purchase_id: number;
    purchase_number: string;
    purchase_date: string;
    supplier_id: number;
    supplier_name: string | null;
    total_amount: number;
    tax_amount: number;
    status: string;
    received_date: string | null;
}

export interface PurchaseReportSummary {
    total_purchases: number;
    total_spending: number;
    avg_purchase_value: number;
}

export interface StockMovementReportItem {
    ledger_id: number;
    created_at: string;
    product_id: number;
    product_name: string | null;
    warehouse_id: number;
    warehouse_name: string | null;
    type: string;
    qty_change: number;
    qty_before: number;
    qty_after: number;
    reference_type: string | null;
    reference_id: number | null;
    note: string | null;
}

export interface StockMovementReportSummary {
    total_movements: number;
    total_units_moved: number;
    total_inbound: number;
    total_outbound: number;
}

export interface InventoryReportResponse {
    items: InventoryReportItem[];
    total: number;
    page: number;
    page_size: number;
    total_pages: number;
}

export interface SalesReportResponse {
    items: SalesReportItem[];
    total: number;
    page: number;
    page_size: number;
    total_pages: number;
    summary: SalesReportSummary;
}

export interface PurchaseReportResponse {
    items: PurchaseReportItem[];
    total: number;
    page: number;
    page_size: number;
    total_pages: number;
    summary: PurchaseReportSummary;
}

export interface StockMovementReportResponse {
    items: StockMovementReportItem[];
    total: number;
    page: number;
    page_size: number;
    total_pages: number;
    summary: StockMovementReportSummary;
}

export interface InventoryReportFilters {
    skip?: number;
    limit?: number;
    warehouse_id?: number;
    category?: string;
    stock_status?: StockStatus;
}

export interface SalesReportFilters {
    skip?: number;
    limit?: number;
    date_from?: string;
    date_to?: string;
    customer_id?: number;
    product_id?: number;
}

export interface PurchaseReportFilters {
    skip?: number;
    limit?: number;
    date_from?: string;
    date_to?: string;
    supplier_id?: number;
    product_id?: number;
}

export interface StockMovementReportFilters {
    skip?: number;
    limit?: number;
    date_from?: string;
    date_to?: string;
    product_id?: number;
    warehouse_id?: number;
    transaction_type?: StockTransactionType;
}

/**
 * Reports API
 */
export const reportsApi = {
    /**
     * Get inventory summary report
     */
    async getInventoryReport(filters: InventoryReportFilters = {}): Promise<InventoryReportResponse> {
        const params = new URLSearchParams();
        if (filters.skip !== undefined) params.append('skip', String(filters.skip));
        if (filters.limit !== undefined) params.append('limit', String(filters.limit));
        if (filters.warehouse_id) params.append('warehouse_id', String(filters.warehouse_id));
        if (filters.category) params.append('category', filters.category);
        if (filters.stock_status) params.append('stock_status', filters.stock_status);

        const url = params.toString() ? `/reports/inventory?${params.toString()}` : '/reports/inventory';
        return http.get<InventoryReportResponse>(url);
    },

    /**
     * Export inventory report to CSV
     */
    async exportInventoryReport(filters: Omit<InventoryReportFilters, 'skip' | 'limit'> = {}): Promise<Blob> {
        const params = new URLSearchParams();
        if (filters.warehouse_id) params.append('warehouse_id', String(filters.warehouse_id));
        if (filters.category) params.append('category', filters.category);
        if (filters.stock_status) params.append('stock_status', filters.stock_status);

        const url = params.toString() ? `/reports/inventory/export?${params.toString()}` : '/reports/inventory/export';
        const response = await fetch(`${import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000/api/v1'}${url}`, {
            headers: {
                'Authorization': `Bearer ${localStorage.getItem('access_token')}`,
            },
        });
        if (!response.ok) throw new Error('Export failed');
        return response.blob();
    },

    /**
     * Get sales report
     */
    async getSalesReport(filters: SalesReportFilters = {}): Promise<SalesReportResponse> {
        const params = new URLSearchParams();
        if (filters.skip !== undefined) params.append('skip', String(filters.skip));
        if (filters.limit !== undefined) params.append('limit', String(filters.limit));
        if (filters.date_from) params.append('date_from', filters.date_from);
        if (filters.date_to) params.append('date_to', filters.date_to);
        if (filters.customer_id) params.append('customer_id', String(filters.customer_id));
        if (filters.product_id) params.append('product_id', String(filters.product_id));

        const url = params.toString() ? `/reports/sales?${params.toString()}` : '/reports/sales';
        return http.get<SalesReportResponse>(url);
    },

    /**
     * Export sales report to CSV
     */
    async exportSalesReport(filters: Omit<SalesReportFilters, 'skip' | 'limit'> = {}): Promise<Blob> {
        const params = new URLSearchParams();
        if (filters.date_from) params.append('date_from', filters.date_from);
        if (filters.date_to) params.append('date_to', filters.date_to);
        if (filters.customer_id) params.append('customer_id', String(filters.customer_id));
        if (filters.product_id) params.append('product_id', String(filters.product_id));

        const url = params.toString() ? `/reports/sales/export?${params.toString()}` : '/reports/sales/export';
        const response = await fetch(`${import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000/api/v1'}${url}`, {
            headers: {
                'Authorization': `Bearer ${localStorage.getItem('access_token')}`,
            },
        });
        if (!response.ok) throw new Error('Export failed');
        return response.blob();
    },

    /**
     * Get purchase report
     */
    async getPurchaseReport(filters: PurchaseReportFilters = {}): Promise<PurchaseReportResponse> {
        const params = new URLSearchParams();
        if (filters.skip !== undefined) params.append('skip', String(filters.skip));
        if (filters.limit !== undefined) params.append('limit', String(filters.limit));
        if (filters.date_from) params.append('date_from', filters.date_from);
        if (filters.date_to) params.append('date_to', filters.date_to);
        if (filters.supplier_id) params.append('supplier_id', String(filters.supplier_id));
        if (filters.product_id) params.append('product_id', String(filters.product_id));

        const url = params.toString() ? `/reports/purchases?${params.toString()}` : '/reports/purchases';
        return http.get<PurchaseReportResponse>(url);
    },

    /**
     * Export purchase report to CSV
     */
    async exportPurchaseReport(filters: Omit<PurchaseReportFilters, 'skip' | 'limit'> = {}): Promise<Blob> {
        const params = new URLSearchParams();
        if (filters.date_from) params.append('date_from', filters.date_from);
        if (filters.date_to) params.append('date_to', filters.date_to);
        if (filters.supplier_id) params.append('supplier_id', String(filters.supplier_id));
        if (filters.product_id) params.append('product_id', String(filters.product_id));

        const url = params.toString() ? `/reports/purchases/export?${params.toString()}` : '/reports/purchases/export';
        const response = await fetch(`${import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000/api/v1'}${url}`, {
            headers: {
                'Authorization': `Bearer ${localStorage.getItem('access_token')}`,
            },
        });
        if (!response.ok) throw new Error('Export failed');
        return response.blob();
    },

    /**
     * Get stock movements report
     */
    async getStockMovementsReport(filters: StockMovementReportFilters = {}): Promise<StockMovementReportResponse> {
        const params = new URLSearchParams();
        if (filters.skip !== undefined) params.append('skip', String(filters.skip));
        if (filters.limit !== undefined) params.append('limit', String(filters.limit));
        if (filters.date_from) params.append('date_from', filters.date_from);
        if (filters.date_to) params.append('date_to', filters.date_to);
        if (filters.product_id) params.append('product_id', String(filters.product_id));
        if (filters.warehouse_id) params.append('warehouse_id', String(filters.warehouse_id));
        if (filters.transaction_type) params.append('transaction_type', filters.transaction_type);

        const url = params.toString() ? `/reports/stock-movements?${params.toString()}` : '/reports/stock-movements';
        return http.get<StockMovementReportResponse>(url);
    },

    /**
     * Export stock movements report to CSV
     */
    async exportStockMovementsReport(filters: Omit<StockMovementReportFilters, 'skip' | 'limit'> = {}): Promise<Blob> {
        const params = new URLSearchParams();
        if (filters.date_from) params.append('date_from', filters.date_from);
        if (filters.date_to) params.append('date_to', filters.date_to);
        if (filters.product_id) params.append('product_id', String(filters.product_id));
        if (filters.warehouse_id) params.append('warehouse_id', String(filters.warehouse_id));
        if (filters.transaction_type) params.append('transaction_type', filters.transaction_type);

        const url = params.toString() ? `/reports/stock-movements/export?${params.toString()}` : '/reports/stock-movements/export';
        const response = await fetch(`${import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000/api/v1'}${url}`, {
            headers: {
                'Authorization': `Bearer ${localStorage.getItem('access_token')}`,
            },
        });
        if (!response.ok) throw new Error('Export failed');
        return response.blob();
    },
};

File: backend/app/api/v1/endpoints/reports.py
"""Reports endpoints."""

from datetime import datetime
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query, Response, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.core.security import check_permission
from app.models.product import StockStatus
from app.models.stock_ledger import StockTransactionType
from app.schemas.reports import (
    InventoryReportResponse,
    PurchaseReportResponse,
    SalesReportResponse,
    StockMovementReportResponse,
)
from app.services.reports_service import ReportsService

router = APIRouter()


@router.get("/inventory", response_model=InventoryReportResponse)
async def get_inventory_report(
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=500),
    warehouse_id: Optional[int] = Query(None),
    category: Optional[str] = Query(None),
    stock_status: Optional[StockStatus] = Query(None),
    db: AsyncSession = Depends(get_db),
    current_user=Depends(check_permission("reports:view")),
):
    """
    Generate inventory summary report.

    Shows current stock levels, values, and status for all products.
    Can be filtered by warehouse, category, or stock status.
    """
    service = ReportsService(db)

    items, total = await service.get_inventory_summary_report(
        warehouse_id=warehouse_id,
        category=category,
        stock_status=stock_status,
        skip=skip,
        limit=limit,
    )

    page = (skip // limit) + 1
    total_pages = (total + limit - 1) // limit

    return InventoryReportResponse(
        items=items,
        total=total,
        page=page,
        page_size=limit,
        total_pages=total_pages,
    )


@router.get("/inventory/export")
async def export_inventory_report(
    warehouse_id: Optional[int] = Query(None),
    category: Optional[str] = Query(None),
    stock_status: Optional[StockStatus] = Query(None),
    db: AsyncSession = Depends(get_db),
    current_user=Depends(check_permission("reports:export")),
):
    """
    Export inventory report to CSV.

    Downloads complete inventory report as CSV file.
    """
    service = ReportsService(db)

    # Get all data (no pagination for export)
    items, _ = await service.get_inventory_summary_report(
        warehouse_id=warehouse_id,
        category=category,
        stock_status=stock_status,
        skip=0,
        limit=10000,  # Large limit for export
    )

    # Generate CSV
    csv_data = ReportsService.generate_csv(
        items,
        fields=[
            "product_id",
            "sku",
            "name",
            "category",
            "quantity",
            "unit",
            "cost_price",
            "selling_price",
            "stock_value",
            "stock_status",
            "reorder_level",
        ],
    )

    # Return as downloadable CSV
    timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
    headers = {
        "Content-Disposition": (
            f"attachment; filename=inventory_report_{timestamp}.csv"
        )
    }
    return Response(content=csv_data, media_type="text/csv", headers=headers)


@router.get("/sales", response_model=SalesReportResponse)
async def get_sales_report(
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=500),
    date_from: Optional[str] = Query(None, description="ISO format date"),
    date_to: Optional[str] = Query(None, description="ISO format date"),
    customer_id: Optional[int] = Query(None),
    product_id: Optional[int] = Query(None),
    db: AsyncSession = Depends(get_db),
    current_user=Depends(check_permission("reports:view")),
):
    """
    Generate sales report with summary statistics.

    Shows completed sales with total revenue, discounts, and order metrics.
    Can be filtered by date range, customer, or product.
    """
    service = ReportsService(db)

    # Parse dates
    parsed_date_from = None
    parsed_date_to = None
    if date_from:
        try:
            parsed_date_from = datetime.fromisoformat(date_from)
        except ValueError:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Invalid date_from format. Use ISO format (YYYY-MM-DD)",
            )
    if date_to:
        try:
            parsed_date_to = datetime.fromisoformat(date_to)
        except ValueError:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Invalid date_to format. Use ISO format (YYYY-MM-DD)",
            )

    items, total, summary = await service.get_sales_report(
        date_from=parsed_date_from,
        date_to=parsed_date_to,
        customer_id=customer_id,
        product_id=product_id,
        skip=skip,
        limit=limit,
    )

    page = (skip // limit) + 1
    total_pages = (total + limit - 1) // limit

    return SalesReportResponse(
        items=items,
        total=total,
        page=page,
        page_size=limit,
        total_pages=total_pages,
        summary=summary,
    )


@router.get("/sales/export")
async def export_sales_report(
    date_from: Optional[str] = Query(None, description="ISO format date"),
    date_to: Optional[str] = Query(None, description="ISO format date"),
    customer_id: Optional[int] = Query(None),
    product_id: Optional[int] = Query(None),
    db: AsyncSession = Depends(get_db),
    current_user=Depends(check_permission("reports:export")),
):
    """
    Export sales report to CSV.

    Downloads complete sales report as CSV file.
    """
    service = ReportsService(db)

    # Parse dates
    parsed_date_from = None
    parsed_date_to = None
    if date_from:
        try:
            parsed_date_from = datetime.fromisoformat(date_from)
        except ValueError:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Invalid date_from format",
            )
    if date_to:
        try:
            parsed_date_to = datetime.fromisoformat(date_to)
        except ValueError:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Invalid date_to format",
            )

    items, _, _ = await service.get_sales_report(
        date_from=parsed_date_from,
        date_to=parsed_date_to,
        customer_id=customer_id,
        product_id=product_id,
        skip=0,
        limit=10000,
    )

    csv_data = ReportsService.generate_csv(
        items,
        fields=[
            "sale_id",
            "invoice_number",
            "sale_date",
            "customer_id",
            "customer_name",
            "total_amount",
            "discount_amount",
            "tax_amount",
            "status",
            "payment_method",
        ],
    )

    timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
    headers = {
        "Content-Disposition": (
            f"attachment; filename=sales_report_{timestamp}.csv"
        )
    }
    return Response(content=csv_data, media_type="text/csv", headers=headers)


@router.get("/purchases", response_model=PurchaseReportResponse)
async def get_purchase_report(
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=500),
    date_from: Optional[str] = Query(None, description="ISO format date"),
    date_to: Optional[str] = Query(None, description="ISO format date"),
    supplier_id: Optional[int] = Query(None),
    product_id: Optional[int] = Query(None),
    db: AsyncSession = Depends(get_db),
    current_user=Depends(check_permission("reports:view")),
):
    """
    Generate purchase report with summary statistics.

    Shows completed purchases with total spending and supplier metrics.
    Can be filtered by date range, supplier, or product.
    """
    service = ReportsService(db)

    # Parse dates
    parsed_date_from = None
    parsed_date_to = None
    if date_from:
        try:
            parsed_date_from = datetime.fromisoformat(date_from)
        except ValueError:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Invalid date_from format. Use ISO format (YYYY-MM-DD)",
            )
    if date_to:
        try:
            parsed_date_to = datetime.fromisoformat(date_to)
        except ValueError:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Invalid date_to format. Use ISO format (YYYY-MM-DD)",
            )

    items, total, summary = await service.get_purchase_report(
        date_from=parsed_date_from,
        date_to=parsed_date_to,
        supplier_id=supplier_id,
        product_id=product_id,
        skip=skip,
        limit=limit,
    )

    page = (skip // limit) + 1
    total_pages = (total + limit - 1) // limit

    return PurchaseReportResponse(
        items=items,
        total=total,
        page=page,
        page_size=limit,
        total_pages=total_pages,
        summary=summary,
    )


@router.get("/purchases/export")
async def export_purchase_report(
    date_from: Optional[str] = Query(None, description="ISO format date"),
    date_to: Optional[str] = Query(None, description="ISO format date"),
    supplier_id: Optional[int] = Query(None),
    product_id: Optional[int] = Query(None),
    db: AsyncSession = Depends(get_db),
    current_user=Depends(check_permission("reports:export")),
):
    """
    Export purchase report to CSV.

    Downloads complete purchase report as CSV file.
    """
    service = ReportsService(db)

    # Parse dates
    parsed_date_from = None
    parsed_date_to = None
    if date_from:
        try:
            parsed_date_from = datetime.fromisoformat(date_from)
        except ValueError:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Invalid date_from format",
            )
    if date_to:
        try:
            parsed_date_to = datetime.fromisoformat(date_to)
        except ValueError:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Invalid date_to format",
            )

    items, _, _ = await service.get_purchase_report(
        date_from=parsed_date_from,
        date_to=parsed_date_to,
        supplier_id=supplier_id,
        product_id=product_id,
        skip=0,
        limit=10000,
    )

    csv_data = ReportsService.generate_csv(
        items,
        fields=[
            "purchase_id",
            "purchase_number",
            "purchase_date",
            "supplier_id",
            "supplier_name",
            "total_amount",
            "tax_amount",
            "status",
            "received_date",
        ],
    )

    timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
    headers = {
        "Content-Disposition": (
            f"attachment; filename=purchase_report_{timestamp}.csv"
        )
    }
    return Response(content=csv_data, media_type="text/csv", headers=headers)


@router.get("/stock-movements", response_model=StockMovementReportResponse)
async def get_stock_movement_report(
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=500),
    date_from: Optional[str] = Query(None, description="ISO format date"),
    date_to: Optional[str] = Query(None, description="ISO format date"),
    product_id: Optional[int] = Query(None),
    warehouse_id: Optional[int] = Query(None),
    transaction_type: Optional[StockTransactionType] = Query(None),
    db: AsyncSession = Depends(get_db),
    current_user=Depends(check_permission("reports:view")),
):
    """
    Generate stock movement report from ledger.

    Shows all stock movements with inbound/outbound statistics.
    Can be filtered by date range, product, warehouse, or transaction type.
    """
    service = ReportsService(db)

    # Parse dates
    parsed_date_from = None
    parsed_date_to = None
    if date_from:
        try:
            parsed_date_from = datetime.fromisoformat(date_from)
        except ValueError:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Invalid date_from format. Use ISO format (YYYY-MM-DD)",
            )
    if date_to:
        try:
            parsed_date_to = datetime.fromisoformat(date_to)
        except ValueError:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Invalid date_to format. Use ISO format (YYYY-MM-DD)",
            )

    items, total, summary = await service.get_stock_movement_report(
        date_from=parsed_date_from,
        date_to=parsed_date_to,
        product_id=product_id,
        warehouse_id=warehouse_id,
        transaction_type=transaction_type,
        skip=skip,
        limit=limit,
    )

    page = (skip // limit) + 1
    total_pages = (total + limit - 1) // limit

    return StockMovementReportResponse(
        items=items,
        total=total,
        page=page,
        page_size=limit,
        total_pages=total_pages,
        summary=summary,
    )


@router.get("/stock-movements/export")
async def export_stock_movement_report(
    date_from: Optional[str] = Query(None, description="ISO format date"),
    date_to: Optional[str] = Query(None, description="ISO format date"),
    product_id: Optional[int] = Query(None),
    warehouse_id: Optional[int] = Query(None),
    transaction_type: Optional[StockTransactionType] = Query(None),
    db: AsyncSession = Depends(get_db),
    current_user=Depends(check_permission("reports:export")),
):
    """
    Export stock movement report to CSV.

    Downloads complete stock movement report as CSV file.
    """
    service = ReportsService(db)

    # Parse dates
    parsed_date_from = None
    parsed_date_to = None
    if date_from:
        try:
            parsed_date_from = datetime.fromisoformat(date_from)
        except ValueError:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Invalid date_from format",
            )
    if date_to:
        try:
            parsed_date_to = datetime.fromisoformat(date_to)
        except ValueError:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Invalid date_to format",
            )

    items, _, _ = await service.get_stock_movement_report(
        date_from=parsed_date_from,
        date_to=parsed_date_to,
        product_id=product_id,
        warehouse_id=warehouse_id,
        transaction_type=transaction_type,
        skip=0,
        limit=10000,
    )

    csv_data = ReportsService.generate_csv(
        items,
        fields=[
            "ledger_id",
            "created_at",
            "product_id",
            "product_name",
            "warehouse_id",
            "warehouse_name",
            "type",
            "qty_change",
            "qty_before",
            "qty_after",
            "reference_type",
            "reference_id",
            "note",
        ],
    )

    timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
    headers = {
        "Content-Disposition": (
            f"attachment; filename=stock_movement_report_{timestamp}.csv"
        )
    }
    return Response(content=csv_data, media_type="text/csv", headers=headers)
